rir/src/compiler/translations/pir_2_rir.cpp

```diff
@@ -0,0 +1,601 @@
+#include "pir_2_rir.h"
+#include "../pir/pir_impl.h"
+#include "../util/cfg.h"
+#include "../util/visitor.h"
+#include "interpreter/runtime.h"
+#include "ir/CodeEditor.h"
+#include "ir/CodeStream.h"
+#include "ir/CodeVerifier.h"
+#include "ir/Optimizer.h"
+#include "utils/FunctionWriter.h"
+
+// #define DEBUGGING
+#define ALLOC_DEBUG 1
+#define PHI_REMOVE_DEBUG 1
+
+#ifdef DEBUGGING
+#define DEBUGCODE(flag, code)                                                  \
+    if (flag)                                                                  \
+    code
+#else
+#define DEBUGCODE(flag, code) /* nothing */
+#endif
+
+namespace rir {
+namespace pir {
+
+namespace {
+
+class Alloc {
+  public:
+    typedef size_t LocalSlotIdx;
+
+    Alloc(Code* code) {
+        BreadthFirstVisitor::run(code->entry, [&](Instruction* instr) {
+            Phi* phi = Phi::Cast(instr);
+            if (phi) {
+                auto slot = allocateLocal(phi);
+                phi->eachArg([&](Value* arg) {
+                    // here the arg must not yet be allocated
+                    if (shouldAllocate(arg))
+                        allocateLocal(arg, slot);
```

Comment:
That is the out-of-SSA transition: you remove the `Phi` by assigning it to a register, and making sure that all inputs are in that same register. Then when executing, the `Phi` is a `noop` because you know that regardless of which path you took, the value is in the `Phi`'s register.